### PR TITLE
bpo-40331: Increase test coverage for the statistics module

### PR DESCRIPTION
**Pull request by tzabal**
https://bugs.python.org/issue40331

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).



## CLA Missing

Our records indicate the following people have not signed the CLA:

@tzabal

For legal reasons we need all the people listed to sign the CLA before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.

If you have recently signed the CLA, please wait at least one business day
before our records are updated.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

File: Lib/statistics.py [hunk 1]
```diff
@@ -267,15 +267,15 @@ def _convert(value, T):
 
 
 def _find_lteq(a, x):
-    'Locate the leftmost value exactly equal to x'
+    """Locate the leftmost value exactly equal to x"""
```
Comment (remilapeyre) [inline Lib/statistics.py]:
```suggestion
    'Locate the leftmost value exactly equal to x'
```
Comestics change are usually refused as they do not add value but cause useless code churn.

File: Lib/statistics.py [hunk 2]
```diff
@@ -267,15 +267,15 @@ def _convert(value, T):
...
     i = bisect_left(a, x)
     if i != len(a) and a[i] == x:
         return i
     raise ValueError
 
 
 def _find_rteq(a, l, x):
-    'Locate the rightmost value exactly equal to x'
+    """Locate the rightmost value exactly equal to x"""
```
Comment (remilapeyre) [inline Lib/statistics.py]:
```suggestion
    'Locate the rightmost value exactly equal to x'
```

File: Lib/test/test_statistics.py [hunk 3]
```diff
@@ -1004,6 +1004,9 @@ def test_nan(self):
             x = statistics._convert(nan, type(nan))
             self.assertTrue(_nan_equal(x, nan))
 
+    def test_raise_type_error(self):
+        self.assertRaises(TypeError, statistics._convert, None, float)
```
Comment (remilapeyre) [inline Lib/test/test_statistics.py]: It's not very good practice to test the internal implementation of a module. Would it be possible to write a test or improve another one to exert this code path without actually depending on the private names?

**Review by remilapeyre — CHANGES_REQUESTED** [review]
Hi @tzabal, thanks for taking the time to improve the tests!

Those changes are not trivial so you will need to sign the CLA and to add a NEWS blurb.

**Comment by tzabal** [comment]
> Hi @tzabal, thanks for taking the time to improve the tests!
> 
> Those changes are not trivial so you will need to sign the CLA and to add a NEWS blurb.

Thank you @remilapeyre for your review! I have signed the CLA an hour before submitting this PR. Sure, I will add a News entry in the next commit.

File: Lib/statistics.py (hunk 1 above)
Comment (tzabal) [reply Lib/statistics.py]: I was a bit hesitant too about this cosmetic change. I will revert it back as you suggested in the next commit.

Let me leave here the reasons that I decided to make the change though:
* I have added unit tests specifically for those two functions, so it is not that I touch code in the codebase that it is not related with the proposed changes of this PR.
* They were the only functions in the `statistics` module that defined their docstrings with single quotes instead of triple double quotes.

File: Lib/statistics.py (hunk 2 above)
Comment (tzabal) [reply Lib/statistics.py]: Will revert this change in the next commit.

File: Lib/test/test_statistics.py (hunk 3 above)
Comment (tzabal) [reply Lib/test/test_statistics.py]: I am not sure that this is possible if we want to hit the 4 missed statements that this PR covers.

The `test_statistics` module defines `TestCase` classes for every defined function of the `statistics` module, even for those defined as private. For instance, the `_convert` function has already 6 unit tests that invoke it directly.

File: Misc/NEWS.d/next/Tests/2020-04-20-00-45-56.bpo-40331.oLPd4g.rst [hunk 4]
```diff
@@ -0,0 +1 @@
+Increase the coverage for the :mod:`statistics` module by 1%.
```
Comment (aeros) [inline Misc/NEWS.d/next/Tests/2020-04-20-00-45-56.bpo-40331.oLPd4g.rst]: In this case, the news entry should include the tests added, not x% coverage. Otherwise, it's not particularly useful to readers.

Alternatively, I think the news entry could be reasonably skipped for this PR. I'll add a skip news label in case you decide to remove it.

Comment (taleinat) [reply Misc/NEWS.d/next/Tests/2020-04-20-00-45-56.bpo-40331.oLPd4g.rst]: I agree, just adding a few tests shouldn't be mentioned in NEWS.

Comment (tzabal) [reply Misc/NEWS.d/next/Tests/2020-04-20-00-45-56.bpo-40331.oLPd4g.rst]: Thank you both @aeros and @taleinat for your input. Checking similar past PRs (e.g., https://github.com/python/cpython/pull/17949), it seems that `skip news` label is the way to go in this case. I will remove it in the next commit.

File: Lib/test/test_statistics.py (hunk 3 above)
Comment (taleinat) [reply Lib/test/test_statistics.py]: Indeed, this is just a small expansion of an existing test suite for these module internals.

IMO given the existing tests this is reasonable and relevant.

Comment (taleinat) [inline Lib/test/test_statistics.py]: I find using `with self.assertRaises():` much more readable, please consider changing this to that style.

```suggestion
        with self.assertRaises(TypeError):
            statistics._convert(None, float)
```

(Also in following new tests in this PR.)

File: Lib/test/test_statistics.py [hunk 5]
```diff
@@ -1454,6 +1496,9 @@ class TestHarmonicMean(NumericTestCase, AverageMixin, UnivariateTypeMixin):
     def setUp(self):
         self.func = statistics.harmonic_mean
 
+    def test_single_value_unsupported_type(self):
```
Comment (taleinat) [inline Lib/test/test_statistics.py]: This should come after the `prepare_` methods.

Also, why test only a an array with a single value of an unsupported type? This would be a good opportunity for other unsupported cases, such as longer arrays, and arrays with values of mixed types, some supported and some not.

**Review by taleinat — CHANGES_REQUESTED** [review]

File: Lib/test/test_statistics.py (hunk 3 above)
Comment (tzabal) [reply Lib/test/test_statistics.py]: Sure, I will change it here, as well as in the `FindLteqTest` and `FindRteqTest` tests, as you suggested.

File: Lib/test/test_statistics.py (hunk 5 above)
Comment (tzabal) [reply Lib/test/test_statistics.py]: You are right, I will put it right after the `test_negative_error` test.

I tested it with a single value of an unsupported type because the statement that explicitly raises this `TypeError` is part of the `elif n == 1` branch, where n is the length of the given list. So a longer list would not trigger this specific `TypeError`, which was the only statement without being covered in this function. [[1]](https://codecov.io/gh/python/cpython/src/8aea4b3605059e243f1827d9328d6fc8d698c0a7/Lib/statistics.py#L404)

If you think that it would be useful to have a test with longer lists (only with unsupported values, both supported/unsupported values) and assert that a `TypeError` is raised, I could add one in the next commit. However, this would be for a `TypeError` that is not explicitly raised by this function.

**Comment by tzabal** [comment]
Thank you @taleinat for the review! I have made the requested changes; please review again.

File: Lib/test/test_statistics.py (hunk 3 above)
Comment (aeros) [reply Lib/test/test_statistics.py]: Sorry to mention it just after you made the change already, but I find that using `assertRaisesRegex()` tends to be a better practice than just `assertRaises()`. It's more significant of an issue for things like deprecation warnings, but it can help quite a bit with ensuring that you're getting the correct exception or warning.

So, it may not be necessary in the above case, but it's worth some consideration.

Comment (taleinat) [reply Lib/test/test_statistics.py]: > I find that using assertRaisesRegex() tends to be a better practice than just assertRaises().

That may be true in general, but I think that in this case it's most appropriate to simply check that a TypeError is thrown.

Comment (aeros) [reply Lib/test/test_statistics.py]: Yeah, looking at the above test a bit more closely, it seems highly unlikely for any other `TypeError` to occur besides the specific one being tested. So, `assertRaises()` is probably fine here.

Comment (tzabal) [reply Lib/test/test_statistics.py]: Thanks for your suggestion @aeros. As you have both already said, it may not be useful in this particular test, but maybe, it would be useful in the `test_single_value_unsupported_type` test (introduced in this PR) because it targets a specific `TypeError` with a custom message, while it is possible for the function to raise other TypeErrors too (through the `_fail_neg` function).

If you don't have any objections, I will go and modify the `test_single_value_unsupported_type` to use the `assertRaisesRegex` instead of the `assertRaises`.

Comment (taleinat) [reply Lib/test/test_statistics.py]: > f you don't have any objections, I will go and modify the `test_single_value_unsupported_type` to use the `assertRaisesRegex` instead of the `assertRaises`.

Go ahead.

File: Lib/test/test_statistics.py (hunk 5 above)
Comment (taleinat) [reply Lib/test/test_statistics.py]: I suggest adding two assertions:
1. multiple string values
2. mixed strings and valid values (i.e. positive numbers).

Comment (tzabal) [reply Lib/test/test_statistics.py]: Thanks for the suggestions @taleinat. I have added a test in the latest commit that asserts those cases.

File: Lib/test/test_statistics.py [hunk 6]
```diff
@@ -1004,6 +1004,10 @@ def test_nan(self):
             x = statistics._convert(nan, type(nan))
             self.assertTrue(_nan_equal(x, nan))
 
+    def test_raise_type_error(self):
```
Comment (taleinat) [inline Lib/test/test_statistics.py]: Let's rename this `test_invalid_input_type()`

File: Lib/test/test_statistics.py [hunk 7]
```diff
@@ -1033,6 +1037,47 @@ def test_error_msg(self):
         self.assertEqual(errmsg, msg)
 
 
+class FindLteqTest(unittest.TestCase):
+    # Test _find_lteq private function.
+
+    def test_raise_value_error(self):
```
Comment (taleinat) [inline Lib/test/test_statistics.py]: Let's also find a better name for this, e.g. `test_invalid_input_values`

File: Lib/test/test_statistics.py [hunk 8]
```diff
@@ -1033,6 +1037,47 @@ def test_error_msg(self):
...
+        for a, x in [
+            ([], 1),        # empty a triggers `i != len(a)`
```
Comment (taleinat) [inline Lib/test/test_statistics.py]: I took a while to think about this, but I've decided that we should remove all of the comments referencing the specific statements that this is intended to exercise. My main reason is that the referenced code will likely change over time, and these comments are very likely to become stale.

File: Lib/test/test_statistics.py [hunk 9]
```diff
@@ -1476,6 +1521,20 @@ def test_negative_error(self):
             with self.subTest(values=values):
                 self.assertRaises(exc, self.func, values)
 
+    def test_single_value_unsupported_type(self):
+        with self.assertRaisesRegex(TypeError, '^unsupported type$'):
```
Comment (taleinat) [inline Lib/test/test_statistics.py]: I don't see the value of this regex here: Would it be wrong if we changed the error message in the future? I prefer that we just check that a `TypeError` is raised.

**Review by taleinat — CHANGES_REQUESTED** [review]
This is nearly ready to go in IMO, just a few final comments.

File: Lib/test/test_statistics.py (hunk 8 above)
Comment (taleinat) [reply Lib/test/test_statistics.py]: The more general reason is that tests must not test the implementation, but only the desired behavior. We might take inspiration for relevant test cases from the code, but the tests should not need to be changed when the code is changed in ways that don't affect its behavior, such as refactoring and optimization.

File: Lib/test/test_statistics.py (hunk 6 above)
Comment (tzabal) [reply Lib/test/test_statistics.py]: Done.

File: Lib/test/test_statistics.py (hunk 7 above)
Comment (tzabal) [reply Lib/test/test_statistics.py]: Done.

File: Lib/test/test_statistics.py (hunk 8 above)
Comment (tzabal) [reply Lib/test/test_statistics.py]: Thank you for describing your reasoning behind this decision. I will remove those comments. I agree that good tests assert the desired behavior and not the actual implementation. I put them only for helping the reader understand why each of them was added, but all of them, trigger the same behavior (throwing the `ValueError`). And as you said, they may become obsolete if the implementation change.

File: Lib/test/test_statistics.py (hunk 9 above)
Comment (tzabal) [reply Lib/test/test_statistics.py]: Done.

**Comment by tzabal** [comment]
Thanks for your review. I have made the requested changes; please review again.

File: Lib/test/test_statistics.py (hunk 8 above)
Comment (taleinat) [reply Lib/test/test_statistics.py]: Comments explaining why, conceptually, each of the test cases should result in such an error, or perhaps which edge-case it is testing, could be welcome. This is true general, and specifically in this case. They're not required though; I'm just adding this note for completeness :)

File: Lib/test/test_statistics.py [hunk 10]
```diff
@@ -1033,6 +1037,47 @@ def test_error_msg(self):
...
+            ([0, 1, 1, 1, 2, 3], 1, 1),
+            ([1, 2, 3, 3, 3], 3, 2)
+        ]:
+            self.assertEqual(expected_i, statistics._find_lteq(a, x))
+
+
+class FindRteqTest(unittest.TestCase):
+    # Test _find_rteq private function.
+
+    def test_raise_value_error(self):
```
Comment (taleinat) [inline Lib/test/test_statistics.py]: I think this test method could also be renamed, similarly to the previous ones.

File: Lib/test/test_statistics.py [hunk 11]
```diff
@@ -1476,6 +1521,20 @@ def test_negative_error(self):
             with self.subTest(values=values):
                 self.assertRaises(exc, self.func, values)
 
+    def test_single_value_unsupported_type(self):
+        with self.assertRaises(TypeError):
+            self.func(['3.14'])
+
+    def test_multiple_values_type_error(self):
```
Comment (taleinat) [inline Lib/test/test_statistics.py]: This could be combined with the previous test method.

**Review by taleinat — APPROVED** [review]
At this point this PR looks good to me.

I'll wait for a week to see if anyone (@rhettinger?) has any comments before this is merged.

File: Lib/test/test_statistics.py (hunk 10 above)
Comment (tzabal) [reply Lib/test/test_statistics.py]: Sorry, I should have changed this too.

File: Lib/test/test_statistics.py (hunk 11 above)
Comment (tzabal) [reply Lib/test/test_statistics.py]: Indeed, now they assert the same behavior, with the same way. I will combine them under a test with name `test_invalid_type_error`.

**Comment by tzabal** [comment]
Thank you for your time reviewing this PR @taleinat, I appreciate your effort.

**Comment by taleinat** [comment]
Thank you, @tzabal, for working on this, receiving review comments with patience and understanding, and iterating on this quickly!

**Review by aeros — APPROVED** [review]
After the recent round of changes, the tests mostly LGTM. 

The only thing I would additionally suggest would be to use [`subTest()`](https://docs.python.org/3.8/library/unittest.html#distinguishing-test-iterations-using-subtests) for the ones that contain 3+ cases: it allows them to fail separately from one another, instead of stopping on the first failure and causing the remaining ones to fail. This makes diagnosing test failures significantly easier when there are multiple cases.

That being said, I think the current changes have significant value either way.

**Comment by tzabal** [comment]
Thank you @aeros for your suggestion. To be honest, I did not know about the existence of `subTest` before. It seems like a nice addition to the `unittest` module. The latest commit utilizes it as you suggested, in tests with 3 or more cases.

**Comment by taleinat** [comment]
Congrats on your first accepted contribution to Python, @tzabal!

It was a pleasure working on this with you and I look forward to doing so again in the future. Please feel free to mention me on issues and PRs as you see fit.

**Comment by tzabal** [comment]
Thanks for reviewing and merging this PR @taleinat! I highly appreciate the time that you devoted in our discussions, and I look forward to collaborating with you again.
